<a href="https://colab.research.google.com/github/ChintaLokesh/Deep_Learning_Neural_Network/blob/master/Module_4_Keras_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn import datasets

In [ ]:
df=pd.read_csv("/content/diabetes.csv")

view the data 

In [ ]:
df.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

check for the NULL values

In [ ]:
df.isna().sum()

Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age                         0
Outcome                     0
dtype: int64

Take the Pregnancies	Glucose	BloodPressure as input parameters

In [ ]:
input_data=df.iloc[:,0:3]

In [ ]:
input_data

Pregnancies  Glucose  BloodPressure
0              6      148             72
1              1       85             66
2              8      183             64
3              1       89             66
4              0      137             40
..           ...      ...            ...
763           10      101             76
764            2      122             70
765            5      121             72
766            1      126             60
767            1       93             70

[768 rows x 3 columns]

Get the target data in to one of the column

In [ ]:
output_variable_train = df.iloc [0:500,8]
output_variable_test = df.iloc [500:767,8]

In [ ]:
output_variable_train

0      1
1      0
2      1
3      0
4      1
      ..
495    0
496    0
497    0
498    1
499    0
Name: Outcome, Length: 500, dtype: int64

In [ ]:
input_data.shape

(768, 3)

In [ ]:
x_train=input_data.iloc[0:500,]

In [ ]:
x_test=input_data.iloc[500:767,]

In [ ]:
x_test.shape

(267, 3)

In [ ]:
x_test

Pregnancies  Glucose  BloodPressure
500            2      117             90
501            3       84             72
502            6        0             68
503            7       94             64
504            3       96             78
..           ...      ...            ...
762            9       89             62
763           10      101             76
764            2      122             70
765            5      121             72
766            1      126             60

[267 rows x 3 columns]

In [ ]:
x_train

Pregnancies  Glucose  BloodPressure
0              6      148             72
1              1       85             66
2              8      183             64
3              1       89             66
4              0      137             40
..           ...      ...            ...
495            6      166             74
496            5      110             68
497            2       81             72
498            7      195             70
499            6      154             74

[500 rows x 3 columns]

In [ ]:
import tensorflow as tf

Create the model

In [ ]:
model=tf.keras.models.Sequential()

Add 8 Nodes and Dense layers

In [ ]:
model.add(tf.keras.layers.Dense(8,input_shape=(3,),activation="relu"))
model.add(tf.keras.layers.Dense(8, activation='relu'))
model.add(tf.keras.layers.Dense(1, activation='sigmoid'))

In [ ]:
model.compile(optimizer=tf.keras.optimizers.SGD(),loss=tf.keras.losses.mae,metrics=["mae"])

Train the model

In [ ]:
model.fit(x_train,output_variable_train,epochs=100,batch_size=10,validation_split=0.1)

Epoch 1/100
45/45 [==============================] - 1s 7ms/step - loss: 0.3824 - mae: 0.3824 - val_loss: 0.2001 - val_mae: 0.2001
Epoch 2/100
45/45 [==============================] - 0s 3ms/step - loss: 0.3822 - mae: 0.3822 - val_loss: 0.2000 - val_mae: 0.2000
Epoch 3/100
45/45 [==============================] - 0s 2ms/step - loss: 0.3822 - mae: 0.3822 - val_loss: 0.2000 - val_mae: 0.2000
Epoch 4/100
45/45 [==============================] - 0s 3ms/step - loss: 0.3822 - mae: 0.3822 - val_loss: 0.2000 - val_mae: 0.2000
Epoch 5/100
45/45 [==============================] - 0s 3ms/step - loss: 0.3822 - mae: 0.3822 - val_loss: 0.2000 - val_mae: 0.2000
Epoch 6/100
45/45 [==============================] - 0s 3ms/step - loss: 0.3822 - mae: 0.3822 - val_loss: 0.2000 - val_mae: 0.2000
Epoch 7/100
45/45 [==============================] - 0s 4ms/step - loss: 0.3822 - mae: 0.3822 - val_loss: 0.2000 - val_mae: 0.2000
Epoch 8/100
45/45 [==============================] - 0s 2ms/step - loss: 0.3822 - m

Evaulate the model

In [ ]:
score = model.evaluate(x_test, output_variable_test, verbose=0)
print(f"score is {score}")
print(f"losss is {score[0]*100}")
print(f"accuracy {score[1]*100}")

score is [0.32209959626197815, 0.32209959626197815]
losss is 32.209959626197815
accuracy 32.209959626197815


From the above metrics , it is clear that model accuracy is 32.2 %

In [ ]:
model.summary()

Model: "sequential_5"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_4 (Dense)             (None, 8)                 32        
                                                                 
 dense_5 (Dense)             (None, 8)                 72        
                                                                 
 dense_6 (Dense)             (None, 1)                 9         
                                                                 
Total params: 113
Trainable params: 113
Non-trainable params: 0
_________________________________________________________________


Create one more model with 16 Nodes

In [ ]:
model=tf.keras.models.Sequential()
model.add(tf.keras.layers.Dense(16,input_shape=(7,),activation="relu"))
model.add(tf.keras.layers.Dense(16,activation="relu"))
model.add(tf.keras.layers.Dense(16,activation="relu"))
model.add(tf.keras.layers.Dense(16,activation="relu"))
model.add(tf.keras.layers.Dense(1,activation="sigmoid"))

In [ ]:
model.summary()

Model: "sequential_8"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_17 (Dense)            (None, 16)                128       
                                                                 
 dense_18 (Dense)            (None, 16)                272       
                                                                 
 dense_19 (Dense)            (None, 16)                272       
                                                                 
 dense_20 (Dense)            (None, 16)                272       
                                                                 
 dense_21 (Dense)            (None, 1)                 17        
                                                                 
Total params: 961
Trainable params: 961
Non-trainable params: 0
_________________________________________________________________


Add optimizer as Adam and loss as Mean Absolute Error 

In [ ]:
model.compile(optimizer=tf.keras.optimizers.Adam(),loss=tf.keras.losses.mae,metrics=["mae"])

split the data into train and test 

In [ ]:
from sklearn.model_selection import train_test_split

In [ ]:
x_train,x_test,y_train,y_test=train_test_split(df.iloc[:,0:7],df.iloc[:,8],test_size=0.2)

Reset the index

In [ ]:
x_test=x_test.reset_index(drop=True)
y_test=y_test.reset_index(drop=True)

In [ ]:
x_test

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              0       93             60              0        0  35.3   
1              2      175             88              0        0  22.9   
2             11      138             76              0        0  33.2   
3              6      103             72             32      190  37.7   
4              3       80             82             31       70  34.2   
..           ...      ...            ...            ...      ...   ...   
149            0      106             70             37      148  39.4   
150            8      110             76              0        0  27.8   
151            1       89             24             19       25  27.8   
152            1       87             78             27       32  34.6   
153            7      103             66             32        0  39.1   

     DiabetesPedigreeFunction  
0                       0.263  
1                       0.326  
2                       0.420  
3                       0.324  
4                       1.292  
..                        ...  
149                     0.605  
150                     0.237  
151                     0.559  
152                     0.101  
153                     0.344  

[154 rows x 7 columns]

In [ ]:
model_history=model.fit(x_train,y_train,epochs=150,batch_size=10)

Epoch 1/150
62/62 [==============================] - 0s 2ms/step - loss: 0.3022 - mae: 0.3022
Epoch 2/150
62/62 [==============================] - 0s 2ms/step - loss: 0.2980 - mae: 0.2980
Epoch 3/150
62/62 [==============================] - 0s 2ms/step - loss: 0.3027 - mae: 0.3027
Epoch 4/150
62/62 [==============================] - 0s 2ms/step - loss: 0.3016 - mae: 0.3016
Epoch 5/150
62/62 [==============================] - 0s 2ms/step - loss: 0.2971 - mae: 0.2971
Epoch 6/150
62/62 [==============================] - 0s 2ms/step - loss: 0.2955 - mae: 0.2955
Epoch 7/150
62/62 [==============================] - 0s 2ms/step - loss: 0.2943 - mae: 0.2943
Epoch 8/150
62/62 [==============================] - 0s 2ms/step - loss: 0.2905 - mae: 0.2905
Epoch 9/150
62/62 [==============================] - 0s 2ms/step - loss: 0.2953 - mae: 0.2953
Epoch 10/150
62/62 [==============================] - 0s 2ms/step - loss: 0.2909 - mae: 0.2909
Epoch 11/150
62/62 [==============================] - 0s 2m

In [ ]:
x_test.shape

(154, 7)

Do predictions

In [187]:
def predict(num,y_test):
  predictions=(model.predict(num)>0.5).astype(int)
  for i in range(10):
    print('Actual  is %d ( and expected is  %d)' % (predictions[i], y_test[i]))

In [188]:
predict(x_test,y_test)

5/5 [==============================] - 0s 3ms/step
Actual  is 0 ( and expected is  0)
Actual  is 1 ( and expected is  0)
Actual  is 1 ( and expected is  0)
Actual  is 0 ( and expected is  0)
Actual  is 0 ( and expected is  1)
Actual  is 0 ( and expected is  0)
Actual  is 0 ( and expected is  1)
Actual  is 0 ( and expected is  0)
Actual  is 0 ( and expected is  1)
Actual  is 0 ( and expected is  0)


Evaulate the model performance

In [189]:
score = model.evaluate(x_test, y_test, verbose=0)
print(f"score is {score}")
print(f"losss is {score[0]*100}")
print(f"accuracy {score[1]*100}")

score is [0.3980957865715027, 0.3980957865715027]
losss is 39.80957865715027
accuracy 39.80957865715027


Model accuracy got increased from 32 to 39